File: kvalifikation/terning2.rb
dices = [
  [  2,
   3,1,6,5,
     4
  ],
  [  3,
   1,2,4,6,
     5
  ],
  [  5,
   6,4,1,3,
     2
  ],
  [  1,
   3,6,2,4,
     5
  ],
  [  4,
   1,6,5,2,
     3
  ]
]

def opposite_index(index)
  case index
  when 0
    5
  when 1
    3
  when 2
    4
  when 3
    1
  when 4
    2
  when 5
    0
  end
end

max_sum = 0

0.upto(5) do |i|
  # Up of fake dice before first to keep DRY
  prev_up = dices.first[opposite_index(i)]
  sum = 0

  dices.each do |dice|
    down_index = dice.index(prev_up)
    up_index = opposite_index(down_index)

    sum += dice.reject { |e| e == dice[up_index] || e == dice[down_index] }.max
    prev_up = dice[up_index]
  end

  sum > max_sum and max_sum = sum
end

p max_sum

File: nationals/kvalifikation/terning2.cc
#include <cstdio>

typedef int dice[6];
dice dices[10000];
int N;
int max_sum;

int opposite_index(int index)
{
  if (index == 0) return 5;
  if (index == 1) return 3;
  if (index == 2) return 4;
  if (index == 3) return 1;
  if (index == 4) return 2;
  return 0; // index = 5, avoid warning
}

void input()
{
  scanf("%d", &N);

  int i;
  for(i = 0; i < N; i++)
    scanf("%d %d %d %d %d %d", &dices[i][0], &dices[i][1], &dices[i][2], &dices[i][3], &dices[i][4], &dices[i][5]);
}

void solve()
{
  int i, j, t;
  int prev_up;
  int sum;
  int largest_side;
  int down_index, up_index;
  int *current_dice;

  // i determines how the dice should be turned
  for (i = 0; i < 6; i++) {
    // turn the first psuedu dice based on i..
    prev_up = dices[0][opposite_index(i)];
    sum = 0;

    // loop through each dice and stack them on top of it
    for (j = 0; j < N; j++) {
      current_dice = dices[j]; // point to current dice

      // find down index, which is where prev_up is
      for (t = 0; t < 6; t++)
        if (current_dice[t] == prev_up) down_index = t;

      up_index = opposite_index(down_index);

      // find maximum of the not up/down sides
      largest_side = 0;
      for (t = 0; t < 6; t++) { // reuse t
        if (t != down_index && t != up_index)
          if (current_dice[t] > largest_side)
            largest_side = current_dice[t];
      }

      // and add this to the sum of the current stack
      sum += largest_side;

      prev_up = current_dice[up_index];
    }

    if (sum > max_sum) max_sum = sum;
  }
}

void output()
{
  printf("%d\n", max_sum);
}

int main(int argc, char * argv[])
{
  input();
  solve();
  output();

  return 0;
}
